---
title: "Mendelian Randomization Guide"
author: "Shea J. Andrews"
date: "2022-11-12"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
setwd('~/Dropbox/Research/UCSF/Snoring-AD-MR/')
```

## Mendelian Randomization
### Load R packages 
```{r packages}
library(tidyverse)    # Data wrangling 
library(TwoSampleMR)  # MR 
library(LDlinkR)      # LD and proxy snps
library(RadialMR)     # Radial MR sensetivity analysis 
library(phenoscanner)
library(MungeSumstats)
# library(tabulizer)

setwd('~/Dropbox/Research/UCSF/Snoring-AD-MR')

# Define column types for summary statistics
coltypes = cols(
  ID = col_character(),
  CHROM = col_double(),
  POS = col_double(),
  REF = col_character(),
  ALT = col_character(),
  AF = col_double(),
  TRAIT = col_character(),
  BETA = col_double(),
  SE = col_double(),
  Z = col_double(),
  P = col_double(),
  N = col_double(),
  OR = col_double(),
  OR_L95 = col_double(),
  OR_U95 = col_double(),
  DIR = col_character(),
  G1000_ID = col_character(),
  G1000_VARIANT = col_character(),
  DBSNP_ID = col_character(),
  DBSNP_VARIANT = col_character(),
  OLD_ID = col_character(),
  OLD_VARIANT = col_character()
)

```

### Exposure dataset
```{r sa_snoring}
## read in and format SA & Snoring GWAS summary stats
## No need to do LD clumping, already done by GWAS
sa_snoring_path = 'data/sasnoring_res.csv'
sa_snoring_ss <- read_csv(sa_snoring_path)

sa_snoring_dat <- sa_snoring_ss %>%
  mutate(Phenotype = "SAsnoring") %>%
  format_data(.,
    type = "exposure",
    snps = NULL,
    header = TRUE,
    phenotype_col = "Phenotype",
    snp_col = "SNP",
    beta_col = "BETA",
    se_col = "SE",
    # eaf_col = "AF",
    effect_allele_col = "A1",
    other_allele_col = "A2",
    pval_col = "P_META",
    chr_col = "CHR",
    pos_col = "BP",
    log_pval = FALSE
) %>%
  as_tibble()

```

```{r snoring}
snoring_path = "~/Dropbox/Research/Data/Summary_Statisitics/Campos2020snor.chrall.CPRA_b37.tsv.gz"
snoring_ss <- read_tsv(snoring_path, comment = "##",  col_types = coltypes, 
                       col_select = c(DBSNP_ID, CHROM, POS, REF, ALT, AF, BETA, SE, Z, P, N))


# Format data to TwoSampleMR format
snoring_exp <- snoring_ss %>%
  mutate(Phenotype = "Snoring") %>%
  format_data(.,
    type = "exposure",
    snps = NULL,
    header = TRUE,
    phenotype_col = "Phenotype",
    snp_col = "DBSNP_ID",
    beta_col = "BETA",
    se_col = "SE",
    eaf_col = "AF",
    effect_allele_col = "ALT",
    other_allele_col = "REF",
    pval_col = "P",
    chr_col = "CHROM",
    pos_col = "POS",
    log_pval = FALSE
) %>%
  as_tibble()

# Perform LD clumping on SNP data, filter SNPs to make it run faster
snoring_clump <- snoring_exp %>% 
  filter(pval.exposure < 0.01) %>%
  clump_data(.,
  clump_kb = 10000,
  clump_r2 = 0.001,
  clump_p1 = 1,
  clump_p2 = 1,
  pop = "EUR"
)

snoring_dat <- filter(snoring_clump, pval.exposure < 5e-8) 
```

```{r snoring_bmi_adj}
snorbmiadj_path <- "~/Dropbox/Research/Data/Summary_Statisitics/Campos2020snorbmiadj.chrall.CPRA_b37.tsv.gz"

snorbmiadj_ss <- read_tsv(snorbmiadj_path, comment = "##",  col_types = coltypes, 
                       col_select = c(DBSNP_ID, CHROM, POS, REF, ALT, AF, BETA, SE, Z, P, N))


# Format data to TwoSampleMR format
snorbmiadj_exp <- snorbmiadj_ss %>%
  mutate(Phenotype = "SnoringBMIadj") %>%
  format_data(.,
    type = "exposure",
    snps = NULL,
    header = TRUE,
    phenotype_col = "Phenotype",
    snp_col = "DBSNP_ID",
    beta_col = "BETA",
    se_col = "SE",
    eaf_col = "AF",
    effect_allele_col = "ALT",
    other_allele_col = "REF",
    pval_col = "P",
    chr_col = "CHROM",
    pos_col = "POS",
    log_pval = FALSE
) %>%
  as_tibble()

# Perform LD clumping on SNP data, filter SNPs to make it run faster
snorbmiadj_clump <- snorbmiadj_exp %>% 
  filter(pval.exposure < 0.01) %>%
  clump_data(.,
  clump_kb = 10000,
  clump_r2 = 0.001,
  clump_p1 = 1,
  clump_p2 = 1,
  pop = "EUR"
)

snorbmiadj_dat <- filter(snorbmiadj_clump, pval.exposure < 5e-8) 

```


## Outcome datasets
Similary, the outcome dataset needs to be converted to the TwoSampleMR format and the exposures SNPs or their LD proxies will be extracted.
```{r outcome}
AD_path = "~/Dropbox/Research/Data/Summary_Statisitics/Kunkle2019load_stage123.chrall.CPRA_b37.tsv.gz"
AD_ss <- read_tsv(AD_path, comment = "##",  col_types = coltypes, 
                       col_select = c(DBSNP_ID, CHROM, POS, REF, ALT, AF, BETA, SE, Z, P, N))

# Format outcome
AD_out <- AD_ss %>%
  mutate(Phenotype = "AD") %>%
  format_data(.,
    type = "outcome",
    snps = NULL,
    header = TRUE,
    phenotype_col = "Phenotype",
    snp_col = "DBSNP_ID",
    beta_col = "BETA",
    se_col = "SE",
    eaf_col = "AF",
    effect_allele_col = "ALT",
    other_allele_col = "REF",
    pval_col = "P",
    samplesize_col = "N",
    z_col = "Z",
    chr_col = "CHROM",
    pos_col = "POS",
    log_pval = FALSE
) %>%
  as_tibble()

```


### LD Proxy SNPs 
```{r get_proxies}
munge_proxies <- function(LDLink_file, outcome, outcome_clump){
  LDLink_file_path <- LDLink_file
  proxy_snps <- read_tsv(LDLink_file_path, skip = 1, col_names = F) %>%
  rename(id = X1, func = X2, proxy_snp = X3, coord = X4, alleles = X5, maf = X6, 
         distance = X7, dprime = X8, rsq = X9, correlated_alleles = X10, FORGEdb = X11, RegulomeDB = X12) %>%
  separate(coord, c('chr', 'pos'), sep = ":") %>%
  mutate(snp = ifelse(id == 1, proxy_snp, NA), 
         chr = str_replace(chr, 'chr', ""), 
         chr = as.numeric(chr), 
         pos = as.numeric(pos)) %>%
  fill(snp, .direction = 'down') %>%
  relocate(snp, .before = proxy_snp) %>%
  dplyr::select(-id, -func, -FORGEdb, -RegulomeDB) %>%
  filter(rsq >= 0.8)

# Munge proxy snp and outcome data
proxy_outcome <- left_join(
  proxy_snps, outcome, by = c("proxy_snp" = "SNP")
) %>%
  separate(correlated_alleles, c("target_a1.outcome", "proxy_a1.outcome", 
                                 "target_a2.outcome", "proxy_a2.outcome"), sep = ",|=") %>%
  filter(!is.na(chr.outcome)) %>%
  arrange(snp, -rsq, abs(distance)) %>%
  group_by(snp) %>%
  slice(1) %>%
  ungroup() %>%
  mutate(
     proxy.outcome = TRUE,
     target_snp.outcome = snp,
     proxy_snp.outcome = proxy_snp, 
  ) %>% 
  mutate(
       new_effect_allele.outcome = case_when(
        proxy_a1.outcome == effect_allele.outcome & proxy_a2.outcome == other_allele.outcome ~ target_a1.outcome,
        proxy_a2.outcome == effect_allele.outcome & proxy_a1.outcome == other_allele.outcome ~ target_a2.outcome,
        TRUE ~ NA_character_
     ), 
      new_other_allele.outcome = case_when(
        proxy_a1.outcome == effect_allele.outcome & proxy_a2.outcome == other_allele.outcome ~ target_a2.outcome,
        proxy_a2.outcome == effect_allele.outcome & proxy_a1.outcome == other_allele.outcome ~ target_a1.outcome,
        TRUE ~ NA_character_
     ), 
     effect_allele.outcome = new_effect_allele.outcome, 
     other_allele.outcome = new_other_allele.outcome
  ) %>%
  dplyr::select(-proxy_snp, -chr, -pos, -alleles, -maf, -distance, -rsq, -dprime,  
         -new_effect_allele.outcome, -new_other_allele.outcome) %>%
  relocate(target_a1.outcome, proxy_a1.outcome, target_a2.outcome, proxy_a2.outcome, .after = proxy_snp.outcome) %>%
  rename(SNP = snp) %>%
  relocate(SNP, .after = samplesize.outcome)

# Merge outcome and proxy outcomes
outcome_dat <- bind_rows(
  outcome_clump, proxy_outcome
) %>% 
  arrange(chr.outcome, pos.outcome)

outcome_dat
}

```

```{r AD_snoring_proxy}
# extract exposure SNPs present in outcome
AD_snoring_clump <- semi_join(
  AD_out, snoring_dat, by = "SNP"
)

# Exposure SNPs not present in outomce
AD_snoring_wo <- anti_join(
  snoring_dat, AD_out, by = "SNP"
)

# Use LDLinkR to identify proxy snps
LDproxy_batch(AD_snoring_wo$SNP, 
        pop = "CEU", 
        r2d = "r2", 
        token = 'a6deee62cc4a', 
        append = TRUE,
        genome_build = "grch37")
system("mv combined_query_snp_list_grch37.txt data/snoring_ad_proxy_snps.txt")

# Munge proxy snp file
AD_snoring_dat <- munge_proxies("data/snoring_ad_proxy_snps.txt", AD_out, AD_snoring_clump)

```

```{r AD_snoring_proxy}
# extract exposure SNPs present in outcome
AD_snorbmiadj_clump <- semi_join(
  AD_out, snorbmiadj_dat, by = "SNP"
)

# Exposure SNPs not present in outomce
AD_snorbmiadj_wo <- anti_join(
  snorbmiadj_dat, AD_out, by = "SNP"
)

# Use LDLinkR to identify proxy snps
LDproxy_batch(AD_snorbmiadj_wo$SNP, 
        pop = "CEU", 
        r2d = "r2", 
        token = 'a6deee62cc4a', 
        append = TRUE,
        genome_build = "grch37")
system("mv combined_query_snp_list_grch37.txt data/snorbmiadj_ad_proxy_snps.txt")

# Munge proxy snp file
AD_snorbmiadj_dat <- munge_proxies("data/snorbmiadj_ad_proxy_snps.txt", AD_out, AD_snorbmiadj_clump)

```

```{r}
# extract exposure SNPs present in outcome
AD_SAsnor <- semi_join(
  AD_out, sa_snoring_dat, by = "SNP"
)

```

## Harmonize Exposure - Outcome Datasets 
```{r harmonize SAsnoring-AD}
SAsnor_AD_mr_dat <- harmonise_data(sa_snoring_dat, AD_SAsnor, action = 2) %>% 
    as_tibble() %>%
    mutate(
    apoe_region = case_when(
      chr.outcome == 19 & between(pos.outcome, 44912079, 45912079) ~ TRUE,
      TRUE ~ FALSE
    ), 
    gws.outcome = ifelse(pval.outcome < 5e-8, TRUE, FALSE), 
    # mr_keep_new = ifelse(Outliers != "Outlier" | apoe_region != FALSE | gws.outcome != FALSE, TRUE, FALSE)
    mr_keep = ifelse(mr_keep == FALSE | apoe_region == TRUE | gws.outcome == TRUE, FALSE, TRUE)
  )

```


```{r harmonize}
# snor_ad_mr_dat <- read_csv('data/snoring_ad_harmonized_data.csv') %>%
# mutate(outcome = 'AD', exposure = 'snoring')
snor_ad_mr_dat <- harmonise_data(snoring_dat, AD_snoring_clump, action = 2) %>% 
    as_tibble() %>%
    mutate(
    apoe_region = case_when(
      chr.outcome == 19 & between(pos.outcome, 44912079, 45912079) ~ TRUE,
      TRUE ~ FALSE
    ), 
    gws.outcome = ifelse(pval.outcome < 5e-8, TRUE, FALSE), 
    # mr_keep_new = ifelse(Outliers != "Outlier" | apoe_region != FALSE | gws.outcome != FALSE, TRUE, FALSE)
    mr_keep = ifelse(mr_keep == FALSE | apoe_region == TRUE | gws.outcome == TRUE, FALSE, TRUE)
  )

```

```{r harmonize}
# snor_ad_mr_dat <- read_csv('data/snoring_ad_harmonized_data.csv') %>%
# mutate(outcome = 'AD', exposure = 'snoring')
snorbmiadj_ad_mr_dat <- harmonise_data(snorbmiadj_dat, AD_snorbmiadj_dat, action = 2) %>% 
    as_tibble() %>%
    mutate(
    apoe_region = case_when(
      chr.outcome == 19 & between(pos.outcome, 44912079, 45912079) ~ TRUE,
      TRUE ~ FALSE
    ), 
    gws.outcome = ifelse(pval.outcome < 5e-8, TRUE, FALSE), 
    # mr_keep_new = ifelse(Outliers != "Outlier" | apoe_region != FALSE | gws.outcome != FALSE, TRUE, FALSE)
    mr_keep = ifelse(mr_keep == FALSE | apoe_region == TRUE | gws.outcome == TRUE, FALSE, TRUE)
  )

```

```{r export_harmonized_data}
mrdat <- bind_rows(
  SAsnor_AD_mr_dat, snor_ad_mr_dat, snorbmiadj_ad_mr_dat
)


write_csv(mrdat, 'data/snoring_ad_harmonized_data.csv')

```

## Perform MR analysis 
```{r MR}
mr_res <- mr(mrdat, method_list = c("mr_ivw_fe", "mr_egger_regression", "mr_weighted_median", "mr_weighted_mode"))

mr_res

```


## Perform Sensentivity analysis
```{r sensetivity}
# Heterogeneity statistics 
het_res <-mr_heterogeneity(mrdat, method_list = c("mr_egger_regression", "mr_ivw"))

# Horizontal pleitropy
plei_res <- mr_pleiotropy_test(mrdat)

# Leave-one-out analysis 
res_loo <- mr_leaveoneout(mrdat, method = mr_ivw_fe) %>% as_tibble()

# Single SNP anlaysis 
res_single <- mr_singlesnp(mrdat, all_method = c("mr_ivw_fe", "mr_egger_regression", "mr_weighted_median", "mr_weighted_mode")) %>% as_tibble()

# Radial MR 
radial_dat <- mrdat %>% filter(mr_keep == T) %>% dat_to_RadialMR()
radial_res_snoring.AD <- ivw_radial(radial_dat$Snoring.AD, alpha = 0.05/nrow(radial_dat$Snoring.AD)) 
radial_res_SAsnoring.AD <- ivw_radial(radial_dat$SAsnoring.AD, alpha = 0.05/nrow(radial_dat$SAsnoring.AD)) 
radial_res_snorbmiadj.AD <- ivw_radial(radial_dat$SnoringBMIadj.AD, alpha = 0.05/nrow(radial_dat$SnoringBMIadj.AD)) 

# Phenoscanner 
phewas_dat <- phenoscanner(snpquery=radial_res$outliers$SNP) 
```

```{r export_results}
write_csv(mr_res, 'results/sleep_ad_mr_results.csv')
write_csv(het_res, 'results/sleep_ad_het_results.csv')
write_csv(plei_res, 'results/sleep_ad_plei_results.csv')
```

## Plots 
```{r plots}
scatter_p <- mr_scatter_plot(mr_res, mrdat)
scatter_p

loo_p <- mr_leaveoneout_plot(res_loo)
loo_p

funnel_p <- mr_funnel_plot(res_single)
funnel_p

radial_snoring.AD_p <- plot_radial(radial_res_snoring.AD, radial_scale = F, show_outliers = T)
radial_snoring.AD_p
radial_snoringbmiadj.AD_p <- plot_radial(radial_res_snorbmiadj.AD, radial_scale = F, show_outliers = T)
radial_snoringbmiadj.AD_p
radial_SAsnoring.AD_p <- plot_radial(radial_res_SAsnoring.AD, radial_scale = F, show_outliers = T)
radial_SAsnoring.AD_p

```

```{r export plots}
# snoring
ggsave('results/plots/mr_scatter_snoringbmiadj.png', plot = scatter_p[[1]], units = 'in', height = 4, width = 9)
ggsave('results/plots/mr_loo_snoringbmiadj.png', plot = loo_p[[1]], units = 'in', height = 9, width = 9)
ggsave('results/plots/mr_funnel_snoringbmiadj.png', plot = funnel_p[[1]], units = 'in', height = 4, width = 9)
ggsave('results/plots/mr_radial_snoringbmiadj.png', plot = radial_snoringbmiadj.AD_p, units = 'in', height = 4, width = 9)

# snoring
ggsave('results/plots/mr_scatter_snoring.png', plot = scatter_p[[2]], units = 'in', height = 4, width = 9)
ggsave('results/plots/mr_loo_snoring.png', plot = loo_p[[2]], units = 'in', height = 9, width = 9)
ggsave('results/plots/mr_funnel_snoring.png', plot = funnel_p[[2]], units = 'in', height = 4, width = 9)
ggsave('results/plots/mr_radial_snoring.png', plot = radial_snoring.AD_p, units = 'in', height = 4, width = 9)

# SAsnoring
ggsave('results/plots/mr_scatter_SAsnoring.png', plot = scatter_p[[2]], units = 'in', height = 4, width = 9)
ggsave('results/plots/mr_loo_SAsnoring.png', plot = loo_p[[2]], units = 'in', height = 9, width = 9)
ggsave('results/plots/mr_funnel_SAsnoring.png', plot = funnel_p[[2]], units = 'in', height = 4, width = 9)
ggsave('results/plots/mr_radial_SAsnoring.png', plot = radial_SAsnoring.AD_p, units = 'in', height = 4, width = 9)

```

## Re-run analysis excluding outliers 
```{r outliers}

## Modify the mrkeep variable to flag variants in the APOE region, that are GWS f
## or the outcome, or are outliers from radial MR for removal
mr_dat_outlier <- mr_dat %>%
  left_join(radial_res$dat) %>%
  mutate(
    apoe_region = case_when(
      chr.outcome == 19 & between(pos.outcome, 44912079, 45912079) ~ TRUE,
      TRUE ~ FALSE
    ), 
    gws.outcome = ifelse(pval.outcome < 5e-8, TRUE, FALSE), 
    # mr_keep_new = ifelse(Outliers != "Outlier" | apoe_region != FALSE | gws.outcome != FALSE, TRUE, FALSE)
     mr_keep = ifelse(mr_keep == FALSE | Outliers == "Outlier" | apoe_region == TRUE | gws.outcome == TRUE, FALSE, TRUE)
  )

## MR analysis 
mr_res_outlier <- mr(mr_dat_outlier, method_list = c("mr_ivw_fe", "mr_egger_regression", "mr_weighted_median", "mr_weighted_mode"))

# Heterogeneity statistics 
 mr_heterogeneity(mr_dat_outlier, method_list = c("mr_egger_regression", "mr_ivw"))

# Horizontal pleitropy
mr_pleiotropy_test(mr_dat_outlier)

# Leave-one-out analysis 
res_loo_outlier <- mr_leaveoneout(mr_dat_outlier, method = mr_ivw_fe) %>% as_tibble()

# Single SNP anlaysis 
res_single_outlier <- mr_singlesnp(mr_dat_outlier, 
                                   all_method = c("mr_ivw_fe", "mr_egger_regression", "mr_weighted_median", "mr_weighted_mode")) %>% 
  as_tibble()

# Radial MR 
radial_dat_outlier <- mr_dat_outlier %>% filter(mr_keep == T) %>% dat_to_RadialMR()
radial_res_outlier <- ivw_radial(radial_dat_outlier$exposure.outcome, alpha = 0.05/nrow(radial_dat_outlier$exposure.outcome)) 

# Plots
scatter_outlier_p <- mr_scatter_plot(mr_res_outlier, mr_dat_outlier)

loo_outlier_p <- mr_leaveoneout_plot(res_loo_outlier)

funnel_outlier_p <- mr_funnel_plot(res_single_outlier)

radial_outlier_p <- plot_radial(radial_res_outlier, radial_scale = F, show_outliers = T)

```

```{r export_results_outliers}
write_csv(mr_res_outlier, 'results/mr_results_outliers.csv')
```

```{r export_outlier_results}
write_csv(mr_res_outlier, 'results/mr_res_outlier.csv')
ggsave('results/plots/mr_scatter_outlier.png', plot = scatter_outlier_p[[1]], units = 'in', height = 4, width = 9)
ggsave('results/plots/mr_loo_outlier.png', plot = loo_outlier_p[[1]], units = 'in', height = 9, width = 9)
ggsave('results/plots/mr_funnel_outlier.png', plot = funnel_outlier_p[[1]], units = 'in', height = 4, width = 9)
ggsave('results/plots/mr_radial_outlier.png', plot = radial_outlier_p, units = 'in', height = 4, width = 9)


```




































