---
title: "MVMR"
output: html_document
date: "2023-01-19"
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = '~/gitcode/Snoring-AD-MR')
```

```{r packages}
library(tidyverse)    # Data wrangling 
library(TwoSampleMR)  # MR 
library(MVMR)
library(RMVMR)
library(LDlinkR)      # LD and proxy snps
library(RadialMR)     # Radial MR sensetivity analysis 
library(phenoscanner)
library(MungeSumstats)
library(cowplot)
# library(MVMRmode)
# library(tabulizer)
`%nin%` = negate(`%in%`)

# Define column types for summary statistics
coltypes = cols(
  ID = col_character(),
  CHROM = col_double(),
  POS = col_double(),
  REF = col_character(),
  ALT = col_character(),
  AF = col_double(),
  TRAIT = col_character(),
  BETA = col_double(),
  SE = col_double(),
  Z = col_double(),
  P = col_double(),
  N = col_double(),
  OR = col_double(),
  OR_L95 = col_double(),
  OR_U95 = col_double(),
  DIR = col_character(),
  G1000_ID = col_character(),
  G1000_VARIANT = col_character(),
  DBSNP_ID = col_character(),
  DBSNP_VARIANT = col_character(),
  OLD_ID = col_character(),
  OLD_VARIANT = col_character()
)

```

```{r get_proxies}
munge_proxies <- function(LDLink_file, outcome, outcome_clump){
  LDLink_file_path <- LDLink_file
  proxy_snps <- read_tsv(LDLink_file_path, skip = 1, col_names = F) %>%
    rename(id = X1, func = X2, proxy_snp = X3, coord = X4, alleles = X5, maf = X6, 
           distance = X7, dprime = X8, rsq = X9, correlated_alleles = X10, FORGEdb = X11, RegulomeDB = X12) %>%
    separate(coord, c('chr', 'pos'), sep = ":") %>%
    mutate(snp = ifelse(id == 1, proxy_snp, NA), 
           chr = str_replace(chr, 'chr', ""), 
           chr = as.numeric(chr), 
           pos = as.numeric(pos)) %>%
    fill(snp, .direction = 'down') %>%
    relocate(snp, .before = proxy_snp) %>%
    dplyr::select(-id, -func, -FORGEdb, -RegulomeDB) %>%
    filter(rsq >= 0.8)

# Munge proxy snp and outcome data
proxy_outcome <- left_join(
    proxy_snps, outcome, by = c("proxy_snp" = "SNP")
  ) %>%
    separate(correlated_alleles, c("target_a1.outcome", "proxy_a1.outcome", 
                                   "target_a2.outcome", "proxy_a2.outcome"), sep = ",|=") %>%
    filter(!is.na(chr.outcome)) %>%
    arrange(snp, -rsq, abs(distance)) %>%
    group_by(snp) %>%
    slice(1) %>%
    ungroup() %>%
    mutate(
       proxy.outcome = TRUE,
       target_snp.outcome = snp,
       proxy_snp.outcome = proxy_snp, 
    ) %>% 
    mutate(
         new_effect_allele.outcome = case_when(
          proxy_a1.outcome == effect_allele.outcome & proxy_a2.outcome == other_allele.outcome ~ target_a1.outcome,
          proxy_a2.outcome == effect_allele.outcome & proxy_a1.outcome == other_allele.outcome ~ target_a2.outcome,
          TRUE ~ NA_character_
       ), 
        new_other_allele.outcome = case_when(
          proxy_a1.outcome == effect_allele.outcome & proxy_a2.outcome == other_allele.outcome ~ target_a2.outcome,
          proxy_a2.outcome == effect_allele.outcome & proxy_a1.outcome == other_allele.outcome ~ target_a1.outcome,
          TRUE ~ NA_character_
       ), 
       effect_allele.outcome = new_effect_allele.outcome, 
       other_allele.outcome = new_other_allele.outcome
    ) %>%
    dplyr::select(-proxy_snp, -chr, -pos, -alleles, -maf, -distance, -rsq, -dprime,  
           -new_effect_allele.outcome, -new_other_allele.outcome) %>%
    relocate(target_a1.outcome, proxy_a1.outcome, target_a2.outcome, proxy_a2.outcome, .after = proxy_snp.outcome) %>%
    rename(SNP = snp) %>%
    relocate(SNP, .after = samplesize.outcome)

# Merge outcome and proxy outcomes
outcome_dat <- bind_rows(
  outcome_clump, proxy_outcome
) %>% 
  arrange(chr.outcome, pos.outcome)

outcome_dat
}

```

## Exposures
```{r exposure-AD}
AD_path = "resources/Kunkle2019load_stage123.chrall.CPRA_b37.tsv.gz"
AD_ss <- read_tsv(AD_path, comment = "##",  col_types = coltypes, 
                       col_select = c(DBSNP_ID, CHROM, POS, REF, ALT, AF, BETA, SE, Z, P, N))

# Format outcome
AD_exp <- AD_ss %>%
  mutate(Phenotype = "AD") %>%
  format_data(.,
    type = "exposure",
    snps = NULL,
    header = TRUE,
    phenotype_col = "Phenotype",
    snp_col = "DBSNP_ID",
    beta_col = "BETA",
    se_col = "SE",
    eaf_col = "AF",
    effect_allele_col = "ALT",
    other_allele_col = "REF",
    pval_col = "P",
    samplesize_col = "N",
    z_col = "Z",
    chr_col = "CHROM",
    pos_col = "POS",
    log_pval = FALSE
) %>%
  as_tibble()


# Perform LD clumping on SNP data, filter SNPs to make it run faster
AD_clump <- AD_exp %>% 
  filter(pval.exposure < 0.01) %>%
  clump_data(.,
  clump_kb = 10000,
  clump_r2 = 0.001,
  clump_p1 = 1,
  clump_p2 = 1,
  pop = "EUR"
)

AD_dat <- filter(AD_clump, pval.exposure < 5e-8) 


```


```{r bmi}
bmi_path <- "resources/Locke2015bmi.chrall.CPRA_b37.tsv.gz"

bmi_ss <- read_tsv(bmi_path, comment = "##",  col_types = coltypes, 
                       col_select = c(DBSNP_ID, CHROM, POS, REF, ALT, AF, BETA, SE, Z, P, N))


# Format data to TwoSampleMR format
bmi_exp <- bmi_ss %>%
  mutate(Phenotype = "BMI") %>%
  format_data(.,
    type = "exposure",
    snps = NULL,
    header = TRUE,
    phenotype_col = "Phenotype",
    snp_col = "DBSNP_ID",
    beta_col = "BETA",
    se_col = "SE",
    eaf_col = "AF",
    effect_allele_col = "ALT",
    other_allele_col = "REF",
    pval_col = "P",
    chr_col = "CHROM",
    pos_col = "POS",
    log_pval = FALSE
) %>%
  as_tibble()

BMI_clump <- bmi_exp %>% 
  filter(pval.exposure < 0.01) %>%
  clump_data(.,
  clump_kb = 10000,
  clump_r2 = 0.001,
  clump_p1 = 1,
  clump_p2 = 1,
  pop = "EUR"
)

BMI_dat <- filter(BMI_clump, pval.exposure < 5e-8) 


## BMI Outcome 
bmi_out <- bmi_ss %>%
  mutate(Phenotype = "BMI") %>%
  format_data(.,
    type = "outcome",
    snps = NULL,
    header = TRUE,
    phenotype_col = "Phenotype",
    snp_col = "DBSNP_ID",
    beta_col = "BETA",
    se_col = "SE",
    eaf_col = "AF",
    effect_allele_col = "ALT",
    other_allele_col = "REF",
    pval_col = "P",
    samplesize_col = "N",
    chr_col = "CHROM",
    pos_col = "POS",
    log_pval = FALSE
) %>%
  as_tibble()


```


## Outcomes
```{r snoring}
snoring_path = "resources/Campos2020snor.chrall.CPRA_b37.tsv.gz"
snoring_ss <- read_tsv(snoring_path, comment = "##",  col_types = coltypes, 
                       col_select = c(DBSNP_ID, CHROM, POS, REF, ALT, AF, BETA, SE, Z, P, N))


# Format data to TwoSampleMR format
snoring_out <- snoring_ss %>%
  mutate(Phenotype = "Snoring") %>%
  format_data(.,
    type = "outcome",
    snps = NULL,
    header = TRUE,
    phenotype_col = "Phenotype",
    snp_col = "DBSNP_ID",
    beta_col = "BETA",
    se_col = "SE",
    eaf_col = "AF",
    effect_allele_col = "ALT",
    other_allele_col = "REF",
    pval_col = "P",
    chr_col = "CHROM",
    pos_col = "POS",
    log_pval = FALSE
) %>%
  as_tibble()

```

```{r snoring_bmi_adj}
snorbmiadj_path <- "resources/Campos2020snorbmiadj.chrall.CPRA_b37.tsv.gz"

snorbmiadj_ss <- read_tsv(snorbmiadj_path, comment = "##",  col_types = coltypes, 
                       col_select = c(DBSNP_ID, CHROM, POS, REF, ALT, AF, BETA, SE, Z, P, N))


# Format data to TwoSampleMR format
snorbmiadj_out <- snorbmiadj_ss %>%
  mutate(Phenotype = "SnoringBMIadj") %>%
  format_data(.,
    type = "outcome",
    snps = NULL,
    header = TRUE,
    phenotype_col = "Phenotype",
    snp_col = "DBSNP_ID",
    beta_col = "BETA",
    se_col = "SE",
    eaf_col = "AF",
    effect_allele_col = "ALT",
    other_allele_col = "REF",
    pval_col = "P",
    chr_col = "CHROM",
    pos_col = "POS",
    log_pval = FALSE
) %>%
  as_tibble()

```

### LD Proxy SNPs 
```{r AD_snoring_proxy}
# extract exposure SNPs present in outcome
snoring_AD_dat <- semi_join(
  snoring_out, AD_dat, by = "SNP"
)

# extract exposure SNPs present in outcome
bmi_AD_dat <- semi_join(
  bmi_out, AD_dat, by = "SNP"
)

# extract exposure SNPs present in outcome
snoringbmiadj_AD_dat <- semi_join(
  snorbmiadj_out, AD_dat, by = "SNP"
)

# extract exposure SNPs present in outcome
snoring_bmi_dat <- semi_join(
  snoring_out, BMI_dat, by = "SNP"
)

# extract exposure SNPs present in outcome
snoringbmiadj_bmi_dat <- semi_join(
  snorbmiadj_out, BMI_dat, by = "SNP"
)

```


## Harmonize Exposure - Outcome Datasets 
```{r harmonize SAsnoring-AD}
Snor_AD_mr_dat <- harmonise_data(AD_dat, snoring_AD_dat, action = 2) %>% 
    as_tibble() %>%
    mutate(
      apoe_region = case_when(
        chr.outcome == 19 & between(pos.outcome, 44912079, 45912079) ~ TRUE,
        TRUE ~ FALSE
      ), 
      gws.outcome = ifelse(pval.outcome < 5e-8, TRUE, FALSE), 
      # mr_keep_new = ifelse(Outliers != "Outlier" | apoe_region != FALSE | gws.outcome != FALSE, TRUE, FALSE)
      # mr_keep = ifelse(mr_keep == FALSE | apoe_region == TRUE | gws.outcome == TRUE, FALSE, TRUE)
      mr_keep = ifelse(mr_keep == FALSE | gws.outcome == TRUE, FALSE, TRUE)
  )

SnorBMIadj_AD_mr_dat <- harmonise_data(AD_dat, snoringbmiadj_AD_dat, action = 2) %>% 
    as_tibble() %>%
    mutate(
      apoe_region = case_when(
        chr.outcome == 19 & between(pos.outcome, 44912079, 45912079) ~ TRUE,
        TRUE ~ FALSE
      ), 
      gws.outcome = ifelse(pval.outcome < 5e-8, TRUE, FALSE), 
      # mr_keep_new = ifelse(Outliers != "Outlier" | apoe_region != FALSE | gws.outcome != FALSE, TRUE, FALSE)
      # mr_keep = ifelse(mr_keep == FALSE | apoe_region == TRUE | gws.outcome == TRUE, FALSE, TRUE)
      mr_keep = ifelse(mr_keep == FALSE | gws.outcome == TRUE, FALSE, TRUE)
  )

Snor_BMI_mr_dat <- harmonise_data(BMI_dat, snoring_bmi_dat, action = 2) %>% 
    as_tibble() %>%
    mutate(
      gws.outcome = ifelse(pval.outcome < 5e-8, TRUE, FALSE), 
      # mr_keep_new = ifelse(Outliers != "Outlier" | apoe_region != FALSE | gws.outcome != FALSE, TRUE, FALSE)
      mr_keep = ifelse(mr_keep == FALSE | gws.outcome == TRUE, FALSE, TRUE)
  )

SnorBMIadj_BMI_mr_dat <- harmonise_data(BMI_dat, snoringbmiadj_bmi_dat, action = 2) %>% 
    as_tibble() %>%
    mutate(
      gws.outcome = ifelse(pval.outcome < 5e-8, TRUE, FALSE), 
      # mr_keep_new = ifelse(Outliers != "Outlier" | apoe_region != FALSE | gws.outcome != FALSE, TRUE, FALSE)
      mr_keep = ifelse(mr_keep == FALSE | gws.outcome == TRUE, FALSE, TRUE)
  )


mrdat_rev <- bind_rows(
  Snor_AD_mr_dat, SnorBMIadj_AD_mr_dat, Snor_BMI_mr_dat, SnorBMIadj_BMI_mr_dat
)


write_csv(mrdat_rev, 'data/ad_bmi_snoring_harmonized_data.csv')

```

## Perform MR analysis 
```{r MR}
# mrdat <- AD_Snor_mr_dat
mr_rev_res <- mr(mrdat_rev, method_list = c("mr_ivw_fe", "mr_ivw", "mr_egger_regression", "mr_weighted_median", "mr_weighted_mode"))

mr_rev_res
generate_odds_ratios(mr_rev_res) %>% group_split(exposure, outcome)
```


## Perform Sensentivity analysis
```{r sensetivity}
# Heterogeneity statistics 
het_res <- mr_heterogeneity(mrdat_rev, method_list = c("mr_egger_regression", "mr_ivw"))

# Horizontal pleitropy
plei_res <- mr_pleiotropy_test(mrdat_rev)

# Leave-one-out analysis 
res_loo <- mr_leaveoneout(mrdat_rev, method = mr_ivw_fe) %>% as_tibble()

# Single SNP anlaysis 
res_single <- mr_singlesnp(mrdat_rev, all_method = c("mr_ivw_fe", "mr_egger_regression", "mr_weighted_median", "mr_weighted_mode")) %>% as_tibble()

# Radial MR 
radial_dat <- mrdat_rev %>% filter(mr_keep == T) %>% dat_to_RadialMR()
radial_res <- map(radial_dat, function(x){
    ivw_radial(x, alpha = 0.05/nrow(x))
  }
)

# Fstat

## F statistic. Burgess et al 2011
f_stat = function(N, K, R){
  f = ((N-K-1) / K) * (R/(1-R))
  f
  }

## Proportion of phenotypic variance explained by SNP 
## https://doi.org/10.1371/journal.pone.0120758.s001
snp.pve <- function(eaf, beta, se, n){
  (2*eaf*(1 - eaf)*beta^2) / (2 * beta * eaf * (1-eaf) + se^2 * 2 * n * eaf * (1-eaf))
}

f_res_rev <- mrdat_rev %>%
  # filter(exposure == "AD" & outcome == "Snoring") %>% 
  group_by(exposure, outcome) %>%
  select(SNP, exposure, outcome, effect_allele.exposure, eaf.exposure, beta.exposure, se.exposure, samplesize.exposure) %>%
  mutate(
    samplesize.exposure = ifelse(exposure == "BMI", 681275, samplesize.exposure),
    eaf.exposure = ifelse(SNP == "rs9270599", 0.6, eaf.exposure),
    pve = snp.pve(eaf.exposure, beta.exposure, se.exposure, samplesize.exposure), 
    f = f_stat(samplesize.exposure, 1, pve),
    # f = abs(beta.exposure)^2 / se.exposure^2
  ) %>% 
  summarise(
    pve = sum(pve, na.rm = T), 
    k = n(), 
    samplesize = max(samplesize.exposure), 
    f = mean(f, na.rm = T),
  )


# Phenoscanner 
phewas_dat <- phenoscanner(snpquery=radial_res$outliers$SNP) 
```

```{r export_results}
write_csv(mr_res, 'results/ad_bmi_sleep_mr_results.csv')
write_csv(het_res, 'results/ad_bmi_sleep_het_results.csv')
write_csv(plei_res, 'results/ad_bmi_sleep_plei_results.csv')
write_csv(f_res_rev, 'results/ad_bmi_f_results.csv')
```


## Plots 
```{r}
## Joint Scatter plot
scatter_rev_p <- mr_scatter_plot(mr_rev_res, mrdat_rev)

joint_scater <- plot_grid(scatter_rev_p[[1]] + theme_bw() + theme(legend.position = 'non', text = element_text(size = 8)), 
          scatter_rev_p[[2]] + theme_bw() + theme(legend.position = 'non', text = element_text(size = 8)), 
          scatter_rev_p[[3]] + theme_bw() + theme(legend.position = 'non', text = element_text(size = 8)), 
          scatter_rev_p[[4]] + theme_bw() + theme(legend.position = 'non', text = element_text(size = 8)))


legend_scatter <- get_legend(
  # create some space to the left of the legend
  scatter_rev_p[[1]] + theme(legend.box.margin = margin(0, 0, 0, 12))
)

# add the legend to the row we made earlier. Give it one-third of 
# the width of one plot (via rel_widths).
scatter_out.p <- plot_grid(joint_scater, legend_scatter, rel_heights = c(3, .4), nrow = 2)

ggsave('results/plots/mr_scatter_ad_bmi_snoring.png', plot = scatter_out.p, units = 'in', height = 4, width = 9)

## Funnel plots
funnel_p <- mr_funnel_plot(res_single)
joint_funnel <- plot_grid(
  funnel_p[[1]] + labs(title = "AD -> SnoringBMIadj") + theme_bw() + theme(legend.position = 'non', text = element_text(size = 8)),
  funnel_p[[2]] + labs(title = "AD -> Snoring") + theme_bw() + theme(legend.position = 'non', text = element_text(size = 8)),
  funnel_p[[3]] + labs(title = "BMI -> SnoringBMIadj") + theme_bw() + theme(legend.position = 'non', text = element_text(size = 8)),
  funnel_p[[4]] + labs(title = "BMI -> Snoring") + theme_bw() + theme(legend.position = 'non', text = element_text(size = 8)))


legend_funnel <- get_legend(
  # create some space to the left of the legend
  scatter_rev_p[[1]] + theme(legend.box.margin = margin(0, 0, 0, 12))
)

# add the legend to the row we made earlier. Give it one-third of 
# the width of one plot (via rel_widths).
funnel_out.p <- plot_grid(joint_funnel, legend_funnel, rel_heights = c(3, .4), nrow = 2)

ggsave('results/plots/mr_funnel_ad_bmi_snoring.png', plot = funnel_out.p, units = 'in', height = 4, width = 9)

## Radial Plots 
radial_p <-  map(radial_res, function(x){
     plot_radial(x, radial_scale = F, show_outliers = F)
  }
)

joint_radial <- plot_grid(
  radial_p[[1]] + labs(title = "AD -> Snoring") + theme_bw() + theme(legend.position = 'non', text = element_text(size = 8)),
  radial_p[[2]] + labs(title = "AD -> SnoringBMIadj") + theme_bw() + theme(legend.position = 'non', text = element_text(size = 8)),
  radial_p[[3]] + labs(title = "BMI -> Snoring") + theme_bw() + theme(legend.position = 'non', text = element_text(size = 8)),
  radial_p[[4]] + labs(title = "BMI -> SnoringBMIadj") + theme_bw() + theme(legend.position = 'non', text = element_text(size = 8))
  )

ggsave('results/plots/mr_radial_ad_bmi_snoring.png', plot = joint_radial, units = 'in', height = 4, width = 9)


```

```{r}
## Scatter plot
ad_snoring_mr_p <- mr_scatter_plot(
  filter(mr_rev_res, exposure == "AD", outcome == "Snoring", method != "Inverse variance weighted"), 
  filter(mrdat_rev, exposure == "AD", outcome == "Snoring")
)

ad_snoring_mr_p_out <- ad_snoring_mr_p[[1]] + 
  labs(tag = "A") + 
  theme_bw() + 
  scale_color_brewer(
    palette = "Set1", 
    labels = c("IVW", "MR-Egger", "WME", "WMBE")) + 
  guides(color = guide_legend(ncol = 1)) + 
  theme(
    legend.position = 'right', 
    text = element_text(size = 8),
    plot.tag = element_text(face = "bold")
    )

## Radial plot
ad_snoring_radial_p_out <- radial_p[[1]] + 
  labs(tag = "C") + 
  theme_bw() + 
  theme(
    legend.position = 'none', 
    text = element_text(size = 8), 
    aspect.ratio=3/5,
    plot.tag = element_text(face = "bold"),
    plot.title = element_blank()
    ) 

## Funnel plot 

ad_snoring_res_single_p <- mr_funnel_plot(
  filter(res_single, exposure == "AD", outcome == "Snoring") 
)

ad_snoring_fun_p_out <- ad_snoring_res_single_p[[1]] + 
  labs(tag = "B") + 
  theme_bw() + 
  scale_color_brewer(
    palette = "Set1", 
    labels = c("IVW", "MR-Egger", "WME", "WMBE")) + 
  theme(
    legend.position = 'bottom',
    text = element_text(size = 8),
    plot.tag = element_text(face = "bold"), 
    )

## Legend
legend_scatter <- get_legend(
  # create some space to the left of the legend
  ad_snoring_mr_p_out + theme(legend.box.margin = margin(0, 0, 0, 12))
)

## Cofficent plot 
ad_snoring_coef_p <- filter(mr_rev_res, exposure == "AD", outcome == "Snoring", method != "Inverse variance weighted") %>% 
  generate_odds_ratios() %>% 
  mutate(
   method = fct_recode(method, "MR-Egger" = "MR Egger", "WME" = "Weighted median", "WMBE" = "Weighted mode", 
                       "IVW" = "Inverse variance weighted (fixed effects)"), 
   method = fct_relevel(method,  "WMBE","WME", "MR-Egger","IVW")
  ) %>%
  ggplot(., aes(x = or, y = method, color = method)) + 
    geom_vline(xintercept = 1, linetype = 2) + 
    geom_point() + 
    geom_errorbar(aes(xmin = or_lci95, xmax = or_uci95), width = 0) + 
    scale_color_manual(values = c("#984EA3", "#4DAF4A", "#377EB8", "#E41A1C")) + 
    labs(tag = "D", x = "OR") + 
    theme_bw() + 
    theme(
        legend.position = 'bottom',
        text = element_text(size = 8),
        plot.tag = element_text(face = "bold"), 
        axis.title.y = element_blank()
  
    )
  
  
    
   

png("docs/ad_snoring_mr.png", width = 7.5, height = 4, units = "in", res = 300)
pageCreate(width = 7.5, height = 4, default.units = "inches")

plotGG(
  plot = ad_snoring_mr_p_out + theme(legend.position = 'none'),
  x = 0, y = 0,
  width = 3, height = 2, just = c("left", "top")
)

plotGG(
  plot = ad_snoring_fun_p_out  + theme(legend.position = 'none'),
  x = 3, y = 0,
  width = 3, height = 2, just = c("left", "top")
)

plotGG(
  plot = legend_scatter,
  x = 6.1, y = 0,
  width = 0.5, height = 2, just = c("left", "top")
)


plotGG(
  plot = ad_snoring_radial_p_out,
  x = 0, y = 2,
  width = 3, height = 2, just = c("left", "top")
)

plotGG(
  plot = ad_snoring_coef_p + theme(legend.position = 'none'),
  x = 3, y = 2,
  width = 3, height = 2, just = c("left", "top")
)


pageGuideHide()
dev.off()
```

```{r}
## Scatter plot
bmi_snoring_mr_p <- mr_scatter_plot(
  filter(mr_rev_res, exposure == "BMI", outcome == "Snoring", method != "Inverse variance weighted"), 
  filter(mrdat_rev, exposure == "BMI", outcome == "Snoring")
)

bmi_snoring_mr_p_out <- bmi_snoring_mr_p[[1]] + 
  labs(tag = "A") + 
  theme_bw() + 
  scale_color_brewer(
    palette = "Set1", 
    labels = c("IVW", "MR-Egger", "WME", "WMBE")) + 
  guides(color = guide_legend(ncol = 1)) + 
  theme(
    legend.position = 'right', 
    text = element_text(size = 8),
    plot.tag = element_text(face = "bold")
    )

## Radial plot
bmi_snoring_radial_p_out <- radial_p[[3]] + 
  labs(tag = "C") + 
  theme_bw() + 
  theme(
    legend.position = 'none', 
    text = element_text(size = 8), 
    aspect.ratio=3/5,
    plot.tag = element_text(face = "bold"),
    plot.title = element_blank()
    ) 

## Funnel plot 

bmi_snoring_res_single_p <- mr_funnel_plot(
  filter(res_single, exposure == "BMI", outcome == "Snoring") 
)

bmi_snoring_fun_p_out <- bmi_snoring_res_single_p[[1]] + 
  labs(tag = "B") + 
  theme_bw() + 
  scale_color_brewer(
    palette = "Set1", 
    labels = c("IVW", "MR-Egger", "WME", "WMBE")) + 
  theme(
    legend.position = 'bottom',
    text = element_text(size = 8),
    plot.tag = element_text(face = "bold"), 
    )

## Legend
bmi_legend_scatter <- get_legend(
  # create some space to the left of the legend
  bmi_snoring_mr_p_out + theme(legend.box.margin = margin(0, 0, 0, 12))
)

## Cofficent plot 
bmi_snoring_coef_p <- filter(mr_rev_res, exposure == "BMI", outcome == "Snoring", method != "Inverse variance weighted") %>% 
  generate_odds_ratios() %>% 
  mutate(
   method = fct_recode(method, "MR-Egger" = "MR Egger", "WME" = "Weighted median", "WMBE" = "Weighted mode", 
                       "IVW" = "Inverse variance weighted (fixed effects)"), 
   method = fct_relevel(method,  "WMBE","WME", "MR-Egger","IVW")
  ) %>%
  ggplot(., aes(x = or, y = method, color = method)) + 
    geom_vline(xintercept = 1, linetype = 2) + 
    geom_point() + 
    geom_errorbar(aes(xmin = or_lci95, xmax = or_uci95), width = 0) + 
    scale_color_manual(values = c("#984EA3", "#4DAF4A", "#377EB8", "#E41A1C")) + 
    labs(tag = "D", x = "OR") + 
    theme_bw() + 
    theme(
        legend.position = 'bottom',
        text = element_text(size = 8),
        plot.tag = element_text(face = "bold"), 
        axis.title.y = element_blank()
  
    )
  

png("docs/bmi_snoring_mr.png", width = 7.5, height = 4, units = "in", res = 300)
pageCreate(width = 7.5, height = 4, default.units = "inches")

plotGG(
  plot = bmi_snoring_mr_p_out + theme(legend.position = 'none'),
  x = 0, y = 0,
  width = 3, height = 2, just = c("left", "top")
)

plotGG(
  plot = bmi_snoring_fun_p_out  + theme(legend.position = 'none'),
  x = 3, y = 0,
  width = 3, height = 2, just = c("left", "top")
)

plotGG(
  plot = bmi_legend_scatter,
  x = 6.1, y = 0,
  width = 0.5, height = 2, just = c("left", "top")
)


plotGG(
  plot = bmi_snoring_radial_p_out,
  x = 0, y = 2,
  width = 3, height = 2, just = c("left", "top")
)

plotGG(
  plot = bmi_snoring_coef_p + theme(legend.position = 'none'),
  x = 3, y = 2,
  width = 3, height = 2, just = c("left", "top")
)


pageGuideHide()
dev.off()
```

## MVMR
```{r}

### AD onto BMI
# extract exposure SNPs present in outcome
AD_bmi_clump <- semi_join(
  bmi_out, AD_dat, by = "SNP"
)

# Exposure SNPs not present in outomce
AD_bmi_wo <- anti_join(
  AD_dat, bmi_out, by = "SNP"
)

# Use LDLinkR to identify proxy snps
LDproxy_batch(AD_bmi_wo$SNP, 
        pop = "CEU", 
        r2d = "r2", 
        token = 'a6deee62cc4a', 
        append = TRUE,
        genome_build = "grch37")
system("mv combined_query_snp_list_grch37.txt data/bmi_ad_proxy_snps.txt")

# Munge proxy snp file
ad_bmi_dat <- munge_proxies("data/bmi_ad_proxy_snps.txt", bmi_out, AD_bmi_clump)

ad_bmi_mr_dat <- harmonise_data(AD_dat, ad_bmi_dat, action = 2) %>% 
    as_tibble() %>%
    bind_rows(filter(Snor_AD_mr_dat, SNP %in% pull(., SNP)))

ad_bmi_res <- mr(ad_bmi_mr_dat, method_list = c("mr_ivw_fe", "mr_egger_regression", "mr_weighted_median", "mr_weighted_mode"))

```

```{r}

mvmr_snps <- c(pull(AD_dat, SNP), pull(BMI_dat, SNP))

bmi_ad_snps <- bind_rows(
  AD_exp %>% filter(pval.exposure < 5e-8), 
  bmi_exp %>% filter(pval.exposure < 5e-8), 
) %>%
  select(SNP, pval.exposure, exposure) %>%
  pivot_wider(names_from = exposure, values_from = pval.exposure) %>%
  mutate(pval.exposure = pmin(AD, BMI, na.rm = T)) 


bmi_ad_clump <- bmi_ad_snps %>% 
  clump_data(.,
  clump_kb = 10000,
  clump_r2 = 0.001,
  clump_p1 = 1,
  clump_p2 = 1,
  pop = "EUR"
)

bmi_ad_mvexp <- bind_rows(
  AD_exp %>% filter(SNP %in% mvmr_snps), 
  bmi_exp %>% filter(SNP %in% mvmr_snps), 
  ad_bmi_dat %>%
      mutate(SNP = target_snp.outcome) %>%
      filter(proxy.outcome == TRUE) %>%
      select(SNP, ends_with('outcome')) %>%
      select(-starts_with('proxy'), -starts_with('target')) %>%
      magrittr::set_colnames(str_replace(colnames(.), 'outcome', 'exposure'))
) %>%
  mutate(
    id.exposure = ifelse(exposure == 'BMI', 'TW76wp', id.exposure)
)

bmi_ad_mvexp %>%
   arrange(chr.exposure, pos.exposure) %>%
   select(SNP, exposure, pval.exposure) %>%
   pivot_wider(names_from = exposure, values_from = pval.exposure) %>%
   count(AD < 5e-8, BMI < 5e-8)
  
```

### TwoSampleMR Package
```{r}
## Harmonize data 
snoring_mvmr_out <- snoring_out %>% filter(SNP %in% bmi_ad_mvexp$SNP)

## TwoSampleMR
# https://mrcieu.github.io/TwoSampleMR/articles/perform_mr.html#multivariable-mr
bmi_ad_mvdat <- mv_harmonise_data(bmi_ad_mvexp, snoring_mvmr_out)

write_rds(bmi_ad_mvdat, 'data/bmi_ad_mvdat.rds.gz')

## MVMR
bmi_ad_res <- mv_multiple(bmi_ad_mvdat)
bmi_ad_res

write_csv(bmi_ad_res$result, 'results/bmi_ad_snoring_mvmr_results.csv')


```

### MVMR package

```{r}
## MVMR
# https://wspiller.github.io/MVMR/articles/MVMR.html
bmi_ad_mvmr_dat <- cbind(bmi_ad_mvdat$exposure_beta, 
        bmi_ad_mvdat$exposure_se, 
        bmi_ad_mvdat$outcome_beta, 
        bmi_ad_mvdat$outcome_se
      ) %>% 
      as.data.frame() %>%
      magrittr::set_colnames(., c("exp1_b", "exp2_b", "exp1_se", "exp2_se", "exp1_p", "exp2_p")) %>%
      rownames_to_column() %>%
      as.matrix()

F.data <- format_mvmr(BXGs = bmi_ad_mvmr_dat[,c(2,3)],
                    BYG = bmi_ad_mvmr_dat[,6],
                    seBXGs = bmi_ad_mvmr_dat[,c(4,5)],
                    seBYG = bmi_ad_mvmr_dat[,7],
                    RSID = bmi_ad_mvmr_dat[,1])
mvmrcovmatrix <- matrix(data = c(1, 0.01, 1, 0.01), nrow = 2, ncol = 2)

sres <- strength_mvmr(r_input = F.data, gencov = 0)
pres <- pleiotropy_mvmr(r_input = F.data, gencov = 0)
res <- ivw_mvmr(r_input = F.data)

qhet_res <- qhet_mvmr(F.data, mvmrcovmatrix, CI = T, iterations = 100)

## RMVMR
F.data_rmvmr <- format_rmvmr(BXGs = bmi_ad_mvmr_dat[,c(2,3)],
                    BYG = bmi_ad_mvmr_dat[,6],
                    seBXGs = bmi_ad_mvmr_dat[,c(4,5)],
                    seBYG = bmi_ad_mvmr_dat[,7],
                    RSID = bmi_ad_mvmr_dat[,1])

res_rmvrm <- ivw_rmvmr(F.data_rmvmr)
sres_rmvrm <- strength_rmvmr(F.data_rmvmr, gencov = 0)
rmvrm_p <- plot_rmvmr(F.data_rmvmr, res_rmvrm)

ggsave('results/plots/rmvmr_bmi_ad_snoring.png', plot = rmvrm_p[[1]], units = 'in', height = 4, width = 9)
```

```{r}
rmvrm_p_out <- rmvrm_p[[1]] + scale_color_manual(
  values = c("#E69F00", "#56B4E9"),  
  labels = c("BMI", "AD")
    ) + 
  labs(color = "Exposure") + 
  theme(
    legend.title = element_blank(), 
    plot.title = element_blank()
  ) 

ggsave('docs/rmvmr_bmi_ad_snoring.png', plot = rmvrm_p_out, units = 'in', height = 4, width = 7.5)


```

```{r}

sres %>%
  pivot_longer(names_to = 'exposure', values_to = 'f', cols = c(exposure1, exposure2)) %>%
  mutate(
    exposure = fct_recode(exposure, 'BMI' = 'exposure1', 'AD' = 'exposure2' ), 
    outcome = 'Snoring'
  ) %>%
  relocate(outcome, .after = exposure) %>%
  write_csv(., 'results/bmi_ad_snoring_sres.csv')

as_tibble(pres) %>% 
  mutate(
    exposure = c("AD | BMI"), 
    outcome = 'Snoring'
  ) %>%
  relocate(exposure, outcome) %>%
  write_csv(., 'results/bmi_ad_snoring_q_res.csv')

```

### MendelianRandomization Package

```{r}

## Format data 
MRMVInputObject <- MendelianRandomization::mr_mvinput(
  bx = bmi_ad_mvdat$exposure_beta, 
  bxse = bmi_ad_mvdat$exposure_se,
  by = bmi_ad_mvdat$outcome_beta,
  byse = bmi_ad_mvdat$outcome_se,
  snps = bmi_ad_mvmr_dat[,1], 
  exposure = c('BMI', 'AD'), 
  outcome = 'Snoring'
  )

## IVW
MRMV_ivw <- MendelianRandomization::mr_mvivw(MRMVInputObject, 
                       model="default", 
                       correl=FALSE, 
                       distribution="normal",
                       alpha = 0.05)

MRMV_ivw_df <- cbind(MRMV_ivw@Exposure, MRMV_ivw@Outcome, MRMV_ivw@Estimate, MRMV_ivw@StdError, MRMV_ivw@Pvalue) %>%
  as_tibble() %>%
  magrittr::set_colnames(c("Exposure", "Outcome", "b", "se", "p")) %>%
  mutate(method = "MRMV-IVW")

## MR-Egger 
MRMV_egger <- MendelianRandomization::mr_mvegger(MRMVInputObject, orientate = 1, correl = FALSE, distribution = "normal", alpha = 0.05)

MRMV_egger_df <- cbind(MRMV_egger@Exposure, MRMV_egger@Outcome, MRMV_egger@Estimate, MRMV_egger@StdError.Est, MRMV_egger@Pvalue.Est) %>%
  as_tibble() %>%
  magrittr::set_colnames(c("Exposure", "Outcome", "b", "se", "p")) %>%
  mutate(method = "MRMV-Egger")

MRMV_egger_intercept <- cbind(MRMV_egger@Exposure, MRMV_egger@Outcome, MRMV_egger@Intercept, MRMV_egger@StdError.Int, MRMV_egger@Pvalue.Int) %>%
  as_tibble() %>%
  magrittr::set_colnames(c("Exposure", "Outcome", "b", "se", "p")) %>%
  slice(1) %>%
  mutate(method = "MRMV-Egger_intecept", 
         Exposure = "AD | BMI")


## WME
MRMV_wme <- MendelianRandomization::mr_mvmedian( MRMVInputObject, distribution = "normal", alpha = 0.05, iterations = 10000, seed = 314159265 )

MRMV_wme_df <- cbind(MRMV_wme@Exposure, MRMV_wme@Outcome, MRMV_wme@Estimate, MRMV_wme@StdError, MRMV_wme@Pvalue) %>%
  as_tibble() %>%
  magrittr::set_colnames(c("Exposure", "Outcome", "b", "se", "p")) %>%
  mutate(method = "MRMV-WME")

## WMBE
MRMV_wbe_res <- MVMRmode::mv_mrmode(Bout = bmi_ad_mvdat$outcome_beta, 
                    Bexp = bmi_ad_mvdat$exposure_beta, 
                    SEout = bmi_ad_mvdat$outcome_se, 
                    SEexp = bmi_ad_mvdat$exposure_se, 
                    # Mode="CM", Psi=0, CIMin = NA, CIMax = NA, CIStep = 0.001, alpha = 0.05, 
                    residual="IVW",
                    Mode="MBE", weighting = "weighted", stderror = "delta", phi = 1,
                    distribution = "normal", iterations = 10000
                    )

MRMV_wbe_df <- as_tibble(MRMV_wbe_res) %>%
  rename(b = Estimate, p = Pvalue, se = StdError) %>%
  mutate(Exposure = c("BMI", "AD"), 
         Outcome = "Snoring", 
         method = 'MRMV_WBE') %>%
  mutate_if(is.double, as.character)


MRMV_out <- 
  bind_rows(
    MRMV_ivw_df, MRMV_egger_df, MRMV_wme_df, MRMV_wbe_df
  ) %>%
  # select(-CILower, -CIUpper, -NvalidSNPs, -Nrange) %>%
  select(-CILower, -CIUpper) %>%
  relocate(method, .after = Outcome) %>%
  mutate_at(vars(b, se, p), as.double)

write_csv(MRMV_out, 'results/bmi_ad_snoring_mvmr_sensetivity_results.csv')
write_csv(MRMV_egger_intercept, 'results/bmi_ad_snoring_mvmr_plei_res.csv')


```






































